### PR TITLE
[Feature] [Lateral-View] support outer combinator of table function

### LINKED ISSUE #9146: [Feature] [Lateral-View] support `outer` combinator of table function
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Now we have some outer table function like `explode_bitmap`, and some non-outer function like `explode`.
I try to make all table functions behave as non-outer, and support `outer` combinator to automaticly build outer version.
example:
```sql
MySQL [test]> select e1 from (select 1 k1) as t lateral view explode_numbers(0) tmp1 as e1;
Empty set (0.009 sec)

MySQL [test]> select e1 from (select 1 k1) as t lateral view explode_numbers_outer(0) tmp1 as e1; 
+------+
| e1   |
+------+
| NULL |
+------+
```

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
# Proposed changes
@HappenLee @cambyzju  please take a look. 
I will update document and unit test later. 
Issue Number: close #9146

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/exprs/table_function/table_function.h [hunk 1]
```diff
@@ -99,6 +99,14 @@ class TableFunction {
     }
 
     bool is_outer() const { return _is_outer; }
+    void set_outer() {
+        if (is_outer()) {
+            return;
+        }
+        _is_outer = true;
+        _fn_name += "_outer";
```
Comment (cambyzju) [inline be/src/exprs/table_function/table_function.h:107]: use suffix_outer

File: be/src/vec/functions/simple_function_factory.h [hunk 2]
```diff
@@ -109,6 +109,13 @@ class SimpleFunctionFactory {
             register_function(Function::name, &Function::create);
     }
 
+    template <class Function>
+    void register_table_function() {
+        function_creators[Function::name] = &createDefaultFunction<Function>;
+        function_creators[std::string(Function::name) + "_outer"] =
```
Comment (cambyzju) [inline be/src/vec/functions/simple_function_factory.h:115]: use suffix_outer

File: be/src/exprs/table_function/table_function_factory.cpp [hunk 3]
```diff
@@ -92,14 +85,32 @@ const std::unordered_map<std::pair<std::string, bool>, std::function<TableFuncti
...
-                {{"explode", true}, VExplodeCreator},
-                {{"explode_outer", true}, VExplodeOuterCreator}}; // namespace doris
+                {{"explode", true}, TableFunctionCreator<vectorized::VExplodeTableFunction> {}}};
+
+Status TableFunctionFactory::get_fn(std::string fn_name, bool is_vectorized, ObjectPool* pool,
+                                    TableFunction** fn) {
+    bool is_outer = false;
+
+    auto match_suffix = [](const std::string& name, const std::string& suffix) -> bool {
+        return name.substr(name.length() - suffix.length()) == suffix;
```
Comment (cambyzju) [inline be/src/exprs/table_function/table_function_factory.cpp:95]: check length at first, maybe name.length() < suffix.length()

File: be/src/exprs/table_function/table_function_factory.cpp [hunk 4]
```diff
@@ -92,14 +85,32 @@ const std::unordered_map<std::pair<std::string, bool>, std::function<TableFuncti
...
+    auto remove_suffix = [](const std::string& name, const std::string& suffix) -> std::string {
+        return name.substr(0, name.length() - suffix.length());
+    };
+
+    if (match_suffix(fn_name, suffix_outer)) {
+        is_outer = true;
+        fn_name = remove_suffix(fn_name, suffix_outer);
+    }
 
-Status TableFunctionFactory::get_fn(const std::string& fn_name, bool is_vectorized,
```
Comment (cambyzju) [inline be/src/exprs/table_function/table_function_factory.cpp:98]: use a local variable, such as std::string search_fn_name for search the function.
Does it looks better than change argument of get_fn function?

File: be/src/exprs/table_function/table_function_factory.cpp (hunk 3 above)
Comment (BiteTheDDDDt) [reply be/src/exprs/table_function/table_function_factory.cpp:95]: I think maybe it's  not necessary to do the check here?
http://www.cplusplus.com/reference/string/string/substr/
`Number of characters to include in the substring (if the string is shorter, as many characters as possible are used).`

**Comment by BiteTheDDDDt** [comment]
@cambyzju fixed

File: be/src/exprs/table_function/table_function.h [hunk 5]
```diff
@@ -30,9 +30,7 @@ namespace doris {
 // Currently, the memory allocated from table function is from malloc directly.
 class TableFunctionState {};
 
-namespace table_function_combinator_suffix {
-static const std::string outer = "_outer";
-}
+const std::string COMBINATOR_SUFFIX_OUTER = "_outer";
```
Comment (HappenLee) [inline be/src/exprs/table_function/table_function.h:33]: constexpr auto

File: be/src/vec/functions/function_fake.cpp [hunk 6]
```diff
@@ -17,18 +17,43 @@
 
 #include "vec/functions/function_fake.h"
 
+#include <boost/metaparse/string.hpp>
+#include <string_view>
+#include <type_traits>
+
 namespace doris::vectorized {
 
+// We can use std::basic_fixed_string with c++20 in future
```
Comment (HappenLee) [inline be/src/vec/functions/function_fake.cpp:26]: in the future

**Review by HappenLee — DISMISSED** [review]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**